src/passes/GlobalRefining.cpp

```diff
@@ -67,9 +67,16 @@ struct GlobalRefining : public Pass {
     // In closed world we cannot change the types of exports, as we might change
     // from a public type to a private that would cause a validation error.
     // TODO We could refine to a type that is still public, however.
+    //
+    // We are also limited in open world: in that mode we must assume that
+    // another module might import our exported globals with the current type
+    // (that type is a contract between them), and in such a case the type of
+    // mutable globals must match precisely (the same rule as for mutable struct
+    // fields in subtypes - the types must match exactly, or else a write in
+    // one place could store a type considered in valid in another place).
     std::unordered_set<Name> unoptimizable;
-    if (getPassOptions().closedWorld) {
-      for (auto* global : ExportUtils::getExportedGlobals(*module)) {
+    for (auto* global : ExportUtils::getExportedGlobals(*module)) {
+      if (getPassOptions().closedWorld || global->mutable_) {
```

Comment:
```suggestion
      if (!getPassOptions().closedWorld || global->mutable_) {
```
Right?

Reply:
I don't think so. Forget mutability for a second, then before we had
```cpp
    if (getPassOptions().closedWorld) {
      for (auto* global : ExportUtils::getExportedGlobals(*module)) {
```
and now we have
```cpp
    for (auto* global : ExportUtils::getExportedGlobals(*module)) {
      if (getPassOptions().closedWorld) {
```
which are equivalent - the check was just moved into the loop (inefficienctly).

Then there is the added check for mutability added (which is why the inefficiency seems worth it, for more concise code here).

And logically, we want to reach new line 80 (which sets unoptimizability) when closed world (so we get there for all exports), or, if open, then when mutable.

Reply:
Wait, but shouldn't the immutable globals be optimizable in closed world?

Reply:
In theory yes, but see pre-existing lines 67-69, we are strict there and avoid all exports, due to public/private issues. There is a TODO to improve that. (This PR changes nothing regarding that.)

Reply:
Ah, I see. Thanks!